# TASK 1 
### Creating a data profile

In [1]:
import pandas as pd

In [2]:
#loading the dataset / reading CSV file
df = pd.read_csv(r"C:\Users\Sai_iluru\OneDrive - UMBC\Desktop\UMBC\733\python notebooks\breast_cancer.csv")
df.head()

Sample code  Clump Thickness  Uniformity of Cell Size  \
0      1000025                5                        1   
1      1002945                5                        4   
2      1015425                3                        1   
3      1016277                6                        8   
4      1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

  Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0           1                3                1        1      2  
1          10                3                2        1      2  
2           2                3                1        1      2  
3           4                3                7        1      2  
4           1                3                1        1      2

In [3]:
import pandas as pd


# Create an empty DataFrame for the data profile
profile_dfs = []

# Iterate through each column and extract information
for col in df.columns:
    data_type = 'int64' if pd.api.types.is_numeric_dtype(df[col]) else 'Object'
    
    if pd.api.types.is_numeric_dtype(df[col]):
        unique_values = df[col].nunique()
        min_val = df[col].min()
        max_val = df[col].max()
    else:
        unique_values = df[col].nunique()
        min_val, max_val = 'NaN','NaN'
    
    profile_col_df = pd.DataFrame({
        'Column Name': [col],
        'Data Type': [data_type],
        'Unique Values': [unique_values],
        'Min': [min_val],
        'Max': [max_val]
    })
    
    profile_dfs.append(profile_col_df)

# Concatenate all DataFrames into a single data profile DataFrame
profile_df = pd.concat(profile_dfs, ignore_index=True)

# Print the data profile table
print(profile_df)


                    Column Name Data Type  Unique Values    Min       Max
0                   Sample code     int64            645  61634  13454352
1               Clump Thickness     int64             10      1        10
2       Uniformity of Cell Size     int64             10      1        10
3      Uniformity of Cell Shape     int64             10      1        10
4             Marginal Adhesion     int64             10      1        10
5   Single Epithelial Cell Size     int64             10      1        10
6                   Bare Nuclei    Object             11    NaN       NaN
7               Bland Chromatin     int64             10      1        10
8               Normal Nucleoli     int64             10      1        10
9                       Mitoses     int64              9      1        10
10                        Class     int64              2      2         4


---------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------
# Task 2
---------------------------------------------------------------------------------------------------------------------------

### Applying the same data profile on the ds_salaries dataset 

In [4]:
#loading the dataset / reading CSV file
df_task2 = pd.read_csv(r"C:\Users\Sai_iluru\OneDrive - UMBC\Desktop\UMBC\733\python notebooks\ds_salaries.csv")
df_task2.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [5]:
df_task2.shape

(3755, 11)

In [6]:


# Create an empty DataFrame for the data profile
profile_dfs = []

# Iterate through each column and extract information
for col in df_task2.columns:
    data_type = 'int64' if pd.api.types.is_numeric_dtype(df_task2[col]) else 'Object'
    
    if pd.api.types.is_numeric_dtype(df_task2[col]):
        unique_values = df_task2[col].nunique()
        min_val = df_task2[col].min()
        max_val = df_task2[col].max()
    else:
        unique_values = df_task2[col].nunique()
        min_val, max_val = 'NaN','NaN'
    
    profile_col_df = pd.DataFrame({
        'Column Name': [col],
        'Data Type': [data_type],
        'Unique Values': [unique_values],
        'Min': [min_val],
        'Max': [max_val]
    })
    
    profile_dfs.append(profile_col_df)

# Concatenate all DataFrames into a single data profile DataFrame
profile_df = pd.concat(profile_dfs, ignore_index=True)

# Print the data profile table
print(profile_df)


           Column Name Data Type  Unique Values   Min       Max
0            work_year     int64              4  2020      2023
1     experience_level    Object              4   NaN       NaN
2      employment_type    Object              4   NaN       NaN
3            job_title    Object             93   NaN       NaN
4               salary     int64            815  6000  30400000
5      salary_currency    Object             20   NaN       NaN
6        salary_in_usd     int64           1035  5132    450000
7   employee_residence    Object             78   NaN       NaN
8         remote_ratio     int64              3     0       100
9     company_location    Object             72   NaN       NaN
10        company_size    Object              3   NaN       NaN


In [7]:
# Check for duplicate rows
duplicates = df_task2.duplicated()

# Count the number of duplicates
num_duplicates = duplicates.sum()

# Display the duplicate rows and count
df_task2[duplicates]

work_year experience_level employment_type           job_title  salary  \
115        2023               SE              FT      Data Scientist  150000   
123        2023               SE              FT  Analytics Engineer  289800   
153        2023               MI              FT       Data Engineer  100000   
154        2023               MI              FT       Data Engineer   70000   
160        2023               SE              FT       Data Engineer  115000   
...         ...              ...             ...                 ...     ...   
3439       2022               MI              FT      Data Scientist   78000   
3440       2022               SE              FT       Data Engineer  135000   
3441       2022               SE              FT       Data Engineer  115000   
3586       2021               MI              FT       Data Engineer  200000   
3709       2021               MI              FT      Data Scientist   76760   

     salary_currency  salary_in_usd employee_residence  remote_ratio  \
115              USD         150000                 US             0   
123              USD         289800                 US             0   
153              USD         100000                 US           100   
154              USD          70000                 US           100   
160              USD         115000                 US             0   
...              ...            ...                ...           ...   
3439             USD          78000                 US           100   
3440             USD         135000                 US           100   
3441             USD         115000                 US           100   
3586             USD         200000                 US           100   
3709             EUR          90734                 DE            50   

     company_location company_size  
115                US            M  
123                US            M  
153                US            M  
154                US            M  
160                US            M  
...               ...          ...  
3439               US            M  
3440               US            M  
3441               US            M  
3586               US            L  
3709               DE            L  

[1171 rows x 11 columns]

---------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------
# Data Report
---------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------

Dear Manager,

I have conducted an initial analysis of the provided dataset and have identified important insights regarding the data's structure. Below is a summary of the key findings: 

## Interesting inferences
 - The data points cover a 5 year data between 2020 to 2023.
 - The dataset has 11 attributes.
 - There are 4 integer valued attributes and 7 attributes that are objects. 
 - The data is a matrix of 3755 rows and 11 columns.
 - It is observed that the data has quite a number of duplicate data points (1171/3755), I think there used to be data from many employees, but perhaps all types of PII info was removed afterwards, leaving us with so many redundant points.
 - It would not be a good idea to remove these points as it will effect the "mean" ( which can be a important factor for various inferences we might draw) of integer attribute types as there cover over 31 % of entire data. 
 
 Hence, I request you to please share any crude version of data from before to verify if the data is really redundant or not.

 - The dataset has various attributes that gives information about job titles, salaries, experience of various candidates along with location of the employee and company and size of the company.
 - I believe we can find interesting relationships between salaries and experience for a job role.
 - We can also draw relations between average salary of a job title in a country.
 - We can normalize the salary column in usd to simply find the countries that provide salary above mean. But this won't give proper answer if we are looking to find ease of living of a employee as we can live a good life with comparitively less salary aat some locations than living with high salary at other locations. we can do this normalization based on the objective we are working for.
 
 ## Possible preprocessing
 
  - After new data is shared I would like to create a unique id for each unique data point based on year the data was collected and the person form whom the data points are collected. This has to be done so that PII of individuals is not compromised with out project objective.
  - Then if there are still any duplicates I will remove them.
  - There are no missing values in this dataset.
  - We can try to normalize the integer attributes like salary by dividing it with minumum value of attribute or we can divide it with mean of the attribute. This has 2 advantages 
        a) Data that we hold in memory will be small.
        b) Sensitive data attributes like salary of a person can be protected it acts like a transformation or encryption.
   - We can convert the attribute with few unique values into numeric values starting like 1,2,3,4...
   
Thank you!!
        

In [8]:
## Pre-processing by dividng with mean
df_task2

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

       salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0       80000             EUR          85847                 ES           100   
1       30000             USD          30000                 US           100   
2       25500             USD          25500                 US           100   
3      175000             USD         175000                 CA           100   
4      120000             USD         120000                 CA           100   
...       ...             ...            ...                ...           ...   
3750   412000             USD         412000                 US           100   
3751   151000             USD         151000                 US           100   
3752   105000             USD         105000                 US           100   
3753   100000             USD         100000                 US           100   
3754  7000000             INR          94665                 IN            50   

     company_location company_size  
0                  ES            L  
1                  US            S  
2                  US            S  
3                  CA            M  
4                  CA            M  
...               ...          ...  
3750               US            L  
3751               US            L  
3752               US            S  
3753               US            L  
3754               IN            L  

[3755 rows x 11 columns]

In [9]:
# Pre-processing the salary usd column
# Column for which mean needs to be calculated and elements divided by the mean
column_to_process = 'salary_in_usd'

# Calculate the mean of the column
column_mean = df_task2[column_to_process].mean()

# Divide each element in the column by the mean
df_task2[column_to_process] = df_task2[column_to_process] / column_mean
df_task2

# as salary is normalized we can remove the salary and salary currency
# Columns to hide (replace with the column names you want to hide)
columns_to_hide = ['salary', 'salary_currency']

# Drop the columns to hide
df_task2_noramlized = df_task2.drop(columns=columns_to_hide)
df_task2_noramlized

work_year experience_level employment_type                 job_title  \
0          2023               SE              FT  Principal Data Scientist   
1          2023               MI              CT               ML Engineer   
2          2023               MI              CT               ML Engineer   
3          2023               SE              FT            Data Scientist   
4          2023               SE              FT            Data Scientist   
...         ...              ...             ...                       ...   
3750       2020               SE              FT            Data Scientist   
3751       2021               MI              FT  Principal Data Scientist   
3752       2020               EN              FT            Data Scientist   
3753       2020               EN              CT     Business Data Analyst   
3754       2021               SE              FT      Data Science Manager   

      salary_in_usd employee_residence  remote_ratio company_location  \
0          0.624022                 ES           100               ES   
1          0.218070                 US           100               US   
2          0.185360                 US           100               US   
3          1.272076                 CA           100               CA   
4          0.872281                 CA           100               CA   
...             ...                ...           ...              ...   
3750       2.994831                 US           100               US   
3751       1.097620                 US           100               US   
3752       0.763246                 US           100               US   
3753       0.726901                 US           100               US   
3754       0.688120                 IN            50               IN   

     company_size  
0               L  
1               S  
2               S  
3               M  
4               M  
...           ...  
3750            L  
3751            L  
3752            S  
3753            L  
3754            L  

[3755 rows x 9 columns]